source/pip/qsharp/_qsharp.py

```diff
@@ -28,6 +35,66 @@
 import sys
 import types
 from time import monotonic
+from dataclasses import make_dataclass, field
+
+
+def lower_python_obj(obj: object, visited: set[object] | None = None) -> Any:
+    if visited is None:
+        visited = set()
+
+    if id(obj) in visited:
+        raise QSharpError("Cannot send circular objects from Python to Q#.")
+
+    visited = visited.copy().add(id(obj))
+
+    # Base case: Primitive types
+    if isinstance(obj, (bool, int, float, complex, str, Pauli, Result)):
+        return obj
+
+    # Recursive case: Tuple
+    if isinstance(obj, tuple):
+        return tuple(lower_python_obj(elt, visited) for elt in obj)
+
+    # Recursive case: Array
+    if isinstance(obj, list):
+        return [lower_python_obj(elt, visited) for elt in obj]
+
+    # Recusive case: Dict
+    if isinstance(obj, dict):
+        return {name: lower_python_obj(val, visited) for name, val in obj.items()}
+
+    # Recursive case: Class
+    if hasattr(obj, "__slots__"):
+        fields = {}
+        for name in obj.__slots__:
+            if name == "__dict__":
+                for name, val in obj.__dict__.items():
+                    fields[name] = lower_python_obj(val, visited)
+            else:
+                val = getattr(obj, name)
+                fields[name] = lower_python_obj(val, visited)
+    elif hasattr(obj, "__dict__"):
+        fields = {
+            name: lower_python_obj(val, visited) for name, val in obj.__dict__.items()
+        }
```

Comment:
Should we raise en error case it an object has `__slots__` and `__dict__` or has a base class that isn't object?